File: HeirloomsApp/app/src/test/kotlin/digital/heirlooms/crypto/VaultCryptoTest.kt
package digital.heirlooms.crypto

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Assert.assertTrue
import org.junit.Test
import java.security.KeyPairGenerator
import java.security.spec.ECGenParameterSpec
import java.security.interfaces.ECPublicKey
import javax.crypto.KeyAgreement

class VaultCryptoTest {

    @Test
    fun `generateMasterKey returns 32 bytes and two calls differ`() {
        val a = VaultCrypto.generateMasterKey()
        val b = VaultCrypto.generateMasterKey()
        assertTrue(a.size == 32)
        assertFalse(a.contentEquals(b))
    }

    @Test
    fun `generateDek returns 32 bytes distinct from master key`() {
        val mk = VaultCrypto.generateMasterKey()
        val dek = VaultCrypto.generateDek()
        assertTrue(dek.size == 32)
        assertFalse(mk.contentEquals(dek))
    }

    @Test
    fun `aesGcmEncrypt decrypt round-trip`() {
        val key = VaultCrypto.generateMasterKey()
        val nonce = VaultCrypto.generateNonce()
        val plaintext = "Hello, Heirlooms vault!".toByteArray()
        val ct = VaultCrypto.aesGcmEncrypt(key, nonce, plaintext)
        val recovered = VaultCrypto.aesGcmDecrypt(key, nonce, ct)
        assertArrayEquals(plaintext, recovered)
    }

    @Test(expected = javax.crypto.AEADBadTagException::class)
    fun `aesGcmDecrypt with wrong key throws`() {
        val key = VaultCrypto.generateMasterKey()
        val wrongKey = VaultCrypto.generateMasterKey()
        val nonce = VaultCrypto.generateNonce()
        val ct = VaultCrypto.aesGcmEncrypt(key, nonce, byteArrayOf(1, 2, 3))
        VaultCrypto.aesGcmDecrypt(wrongKey, nonce, ct)
    }

    @Test
    fun `buildSymmetricEnvelope has correct structure`() {
        val algId = VaultCrypto.ALG_AES256GCM_V1
        val nonce = ByteArray(12) { it.toByte() }
        val ct = ByteArray(32) { (it + 100).toByte() }
        val env = VaultCrypto.buildSymmetricEnvelope(algId, nonce, ct)

        val version = env[0].toInt() and 0xFF
        val algLen = env[1].toInt() and 0xFF
        val algBytes = env.copyOfRange(2, 2 + algLen)

        assertTrue(version == 1)
        assertTrue(String(algBytes) == algId)
    }

    @Test
    fun `encryptSymmetric then decryptSymmetric round-trip`() {
        val key = VaultCrypto.generateMasterKey()
        val plaintext = "secret data".toByteArray()
        val envelope = VaultCrypto.encryptSymmetric(VaultCrypto.ALG_AES256GCM_V1, key, plaintext)
        val recovered = VaultCrypto.decryptSymmetric(envelope, key)
        assertArrayEquals(plaintext, recovered)
    }

    @Test
    fun `wrapDekUnderMasterKey unwrapDekWithMasterKey round-trip`() {
        val masterKey = VaultCrypto.generateMasterKey()
        val dek = VaultCrypto.generateDek()
        val wrapped = VaultCrypto.wrapDekUnderMasterKey(dek, masterKey)
        val recovered = VaultCrypto.unwrapDekWithMasterKey(wrapped, masterKey)
        assertArrayEquals(dek, recovered)
    }

    @Test(expected = Exception::class)
    fun `decryptSymmetric with wrong key throws`() {
        val key = VaultCrypto.generateMasterKey()
        val wrongKey = VaultCrypto.generateMasterKey()
        val envelope = VaultCrypto.encryptSymmetric(VaultCrypto.ALG_AES256GCM_V1, key, byteArrayOf(1, 2, 3))
        VaultCrypto.decryptSymmetric(envelope, wrongKey)
    }

    @Test
    fun `hkdf is deterministic for same inputs and differs for different IKMs`() {
        val ikm1 = ByteArray(32) { 0xAB.toByte() }
        val ikm2 = ByteArray(32) { 0xCD.toByte() }
        val info = "heirlooms-v1".toByteArray()
        val out1a = VaultCrypto.hkdf(ikm1, info = info)
        val out1b = VaultCrypto.hkdf(ikm1, info = info)
        val out2 = VaultCrypto.hkdf(ikm2, info = info)
        assertTrue(out1a.size == 32)
        assertArrayEquals(out1a, out1b)
        assertFalse(out1a.contentEquals(out2))
    }

    @Test
    fun `wrapMasterKeyWithPassphrase returns valid envelope`() {
        val masterKey = VaultCrypto.generateMasterKey()
        val passphrase = "correct-horse-battery-staple".toCharArray()
        val wrap = VaultCrypto.wrapMasterKeyWithPassphrase(masterKey, passphrase)
        assertTrue(wrap.envelope.isNotEmpty())
        assertTrue(wrap.salt.size == 16)
        assertTrue(wrap.envelope[0].toInt() and 0xFF == 1)  // version byte
    }

    @Test
    fun `wrapMasterKeyWithPassphrase unwrapMasterKeyWithPassphrase round-trip`() {
        val masterKey = VaultCrypto.generateMasterKey()
        val passphrase = "correct-horse-battery-staple".toCharArray()
        val wrap = VaultCrypto.wrapMasterKeyWithPassphrase(masterKey, passphrase)
        val recovered = VaultCrypto.unwrapMasterKeyWithPassphrase(
            wrap.envelope, passphrase, wrap.salt, wrap.params
        )
        assertArrayEquals(masterKey, recovered)
    }

    @Test(expected = Exception::class)
    fun `unwrapMasterKeyWithPassphrase wrong passphrase throws`() {
        val masterKey = VaultCrypto.generateMasterKey()
        val passphrase = "correct-horse-battery-staple".toCharArray()
        val wrongPassphrase = "wrong-passphrase".toCharArray()
        val wrap = VaultCrypto.wrapMasterKeyWithPassphrase(masterKey, passphrase)
        VaultCrypto.unwrapMasterKeyWithPassphrase(wrap.envelope, wrongPassphrase, wrap.salt, wrap.params)
    }

    @Test
    fun `buildAsymmetricEnvelope parseAsymmetricEnvelope round-trip`() {
        val algId = VaultCrypto.ALG_P256_ECDH_HKDF_V1
        val ephPub = ByteArray(65) { (it + 4).toByte() }.also { it[0] = 0x04 }
        val nonce = ByteArray(12) { it.toByte() }
        val ct = ByteArray(48) { it.toByte() }
        val env = VaultCrypto.buildAsymmetricEnvelope(algId, ephPub, nonce, ct)
        val parsed = VaultCrypto.parseAsymmetricEnvelope(env)
        assertTrue(parsed.algorithmId == algId)
        assertArrayEquals(ephPub, parsed.ephemeralPubkeyBytes)
        assertArrayEquals(nonce, parsed.nonce)
        assertArrayEquals(ct, parsed.ciphertextWithTag)
    }

    @Test
    fun `asymmetric envelope ECDH round-trip encrypts and decrypts master key`() {
        // Simulate device keypair (static) in software
        val kpg = KeyPairGenerator.getInstance("EC")
        kpg.initialize(ECGenParameterSpec("secp256r1"))
        val deviceKp = kpg.generateKeyPair()
        val devicePub = deviceKp.public as ECPublicKey

        // Encode device public key as SEC1 uncompressed
        val x = devicePub.w.affineX.toByteArray().let { if (it.size == 33) it.copyOfRange(1, 33) else it.padTo(32) }
        val y = devicePub.w.affineY.toByteArray().let { if (it.size == 33) it.copyOfRange(1, 33) else it.padTo(32) }
        val devicePubSec1 = byteArrayOf(0x04) + x + y

        // Wrap master key under device public key (ephemeral)
        val masterKey = VaultCrypto.generateMasterKey()
        val ephKpg = KeyPairGenerator.getInstance("EC")
        ephKpg.initialize(ECGenParameterSpec("secp256r1"))
        val ephKp = ephKpg.generateKeyPair()
        val ephPub = ephKp.public as ECPublicKey
        val ephPubX = ephPub.w.affineX.toByteArray().let { if (it.size == 33) it.copyOfRange(1, 33) else it.padTo(32) }
        val ephPubY = ephPub.w.affineY.toByteArray().let { if (it.size == 33) it.copyOfRange(1, 33) else it.padTo(32) }
        val ephPubSec1 = byteArrayOf(0x04) + ephPubX + ephPubY

        // ECDH: eph_private × device_public → shared_secret (same as device_private × eph_public)
        val ka1 = KeyAgreement.getInstance("ECDH")
        ka1.init(ephKp.private)
        ka1.doPhase(deviceKp.public, true)
        val sharedSecret1 = ka1.generateSecret()

        val kek = VaultCrypto.hkdf(sharedSecret1, info = "heirlooms-v1".toByteArray())
        val nonce = VaultCrypto.generateNonce()
        val ct = VaultCrypto.aesGcmEncrypt(kek, nonce, masterKey)
        val envelope = VaultCrypto.buildAsymmetricEnvelope(VaultCrypto.ALG_P256_ECDH_HKDF_V1, ephPubSec1, nonce, ct)

        // Unwrap using device private key + ephemeral public from envelope
        val parsed = VaultCrypto.parseAsymmetricEnvelope(envelope)
        val ephPubFromEnvelope = decodeEcSec1(parsed.ephemeralPubkeyBytes)
        val ka2 = KeyAgreement.getInstance("ECDH")
        ka2.init(deviceKp.private)
        ka2.doPhase(ephPubFromEnvelope, true)
        val sharedSecret2 = ka2.generateSecret()
        val kek2 = VaultCrypto.hkdf(sharedSecret2, info = "heirlooms-v1".toByteArray())
        val recovered = VaultCrypto.aesGcmDecrypt(kek2, parsed.nonce, parsed.ciphertextWithTag)

        assertArrayEquals(masterKey, recovered)
    }

    private fun decodeEcSec1(bytes: ByteArray): java.security.PublicKey {
        val x = java.math.BigInteger(1, bytes.copyOfRange(1, 33))
        val y = java.math.BigInteger(1, bytes.copyOfRange(33, 65))
        val params = java.security.AlgorithmParameters.getInstance("EC")
        params.init(ECGenParameterSpec("secp256r1"))
        val curveParams = params.getParameterSpec(java.security.spec.ECParameterSpec::class.java)
        val spec = java.security.spec.ECPublicKeySpec(java.security.spec.ECPoint(x, y), curveParams)
        return java.security.KeyFactory.getInstance("EC").generatePublic(spec)
    }

    private fun ByteArray.padTo(size: Int): ByteArray =
        if (this.size >= size) this else ByteArray(size - this.size) + this
}

File: HeirloomsWeb/src/test/vaultCrypto.test.js
import { describe, it, expect } from 'vitest'

// vitest's toEqual doesn't always deep-compare Uint8Array correctly across
// different ArrayBuffer origins; convert to plain arrays for comparisons.
function bytes(u) { return Array.from(u) }
import {
  generateMasterKey, generateDek, generateNonce, generateSalt,
  aesGcmEncrypt, aesGcmDecrypt, aesGcmEncryptWithAad,
  buildSymmetricEnvelope, parseSymmetricEnvelope,
  encryptSymmetric, decryptSymmetric, decryptStreamingContent,
  wrapDekUnderMasterKey, unwrapDekWithMasterKey,
  hkdf,
  wrapMasterKeyWithPassphrase, unwrapMasterKeyWithPassphrase,
  buildAsymmetricEnvelope, parseAsymmetricEnvelope,
  wrapMasterKeyForDevice,
  ALG_AES256GCM_V1,
  DEFAULT_ARGON2_PARAMS,
} from '../crypto/vaultCrypto'

// Use minimal Argon2 params so tests run fast.
const FAST_PARAMS = { m: 64, t: 1, p: 1 }

describe('vaultCrypto', () => {
  it('1 — generateMasterKey returns 32 bytes, two calls differ', () => {
    const a = generateMasterKey()
    const b = generateMasterKey()
    expect(a).toHaveLength(32)
    expect(bytes(a)).not.toEqual(bytes(b))
  })

  it('2 — generateDek returns 32 bytes, differs from master key', () => {
    const mk = generateMasterKey()
    const dek = generateDek()
    expect(dek).toHaveLength(32)
    expect(bytes(dek)).not.toEqual(bytes(mk))
  })

  it('3 — aesGcmEncrypt / aesGcmDecrypt round-trip', async () => {
    const key = generateMasterKey()
    const nonce = generateNonce()
    const plaintext = new TextEncoder().encode('hello heirlooms')
    const ct = await aesGcmEncrypt(key, nonce, plaintext)
    const recovered = await aesGcmDecrypt(key, nonce, ct)
    expect(bytes(recovered)).toEqual(bytes(plaintext))
  })

  it('4 — aesGcmDecrypt with wrong key throws', async () => {
    const key = generateMasterKey()
    const nonce = generateNonce()
    const ct = await aesGcmEncrypt(key, nonce, new Uint8Array([1, 2, 3]))
    const wrongKey = generateMasterKey()
    await expect(aesGcmDecrypt(wrongKey, nonce, ct)).rejects.toThrow()
  })

  it('5 — buildSymmetricEnvelope byte layout', () => {
    const alg = 'aes256gcm-v1'
    const algBytes = new TextEncoder().encode(alg)
    const nonce = generateNonce()
    const ct = new Uint8Array([10, 20, 30])
    const env = buildSymmetricEnvelope(alg, nonce, ct)

    expect(env[0]).toBe(0x01)                          // version
    expect(env[1]).toBe(algBytes.length)               // alg_id_len
    const algSlice = env.slice(2, 2 + algBytes.length)
    expect(bytes(algSlice)).toEqual(bytes(algBytes))   // alg_id
    const nonceSlice = env.slice(2 + algBytes.length, 2 + algBytes.length + 12)
    expect(bytes(nonceSlice)).toEqual(bytes(nonce))    // nonce
    const ctSlice = env.slice(2 + algBytes.length + 12)
    expect(bytes(ctSlice)).toEqual(bytes(ct))          // ciphertext
  })

  it('6 — encryptSymmetric / decryptSymmetric round-trip', async () => {
    const key = generateMasterKey()
    const msg = new TextEncoder().encode('round trip test')
    const envelope = await encryptSymmetric(ALG_AES256GCM_V1, key, msg)
    const recovered = await decryptSymmetric(envelope, key)
    expect(bytes(recovered)).toEqual(bytes(msg))
  })

  it('7 — wrapDekUnderMasterKey / unwrapDekWithMasterKey round-trip', async () => {
    const masterKey = generateMasterKey()
    const dek = generateDek()
    const wrapped = await wrapDekUnderMasterKey(dek, masterKey)
    const recovered = await unwrapDekWithMasterKey(wrapped, masterKey)
    expect(bytes(recovered)).toEqual(bytes(dek))
  })

  it('8 — decryptSymmetric with wrong key throws', async () => {
    const key = generateMasterKey()
    const envelope = await encryptSymmetric(ALG_AES256GCM_V1, key, new Uint8Array([1]))
    const wrongKey = generateMasterKey()
    await expect(decryptSymmetric(envelope, wrongKey)).rejects.toThrow()
  })

  it('9 — hkdf is deterministic for same inputs, differs for different IKMs', async () => {
    const ikm = generateMasterKey()
    const info = new TextEncoder().encode('test-info')
    const out1 = await hkdf(ikm, null, info)
    const out2 = await hkdf(ikm, null, info)
    expect(bytes(out1)).toEqual(bytes(out2))
    expect(out1).toHaveLength(32)

    const ikm2 = generateMasterKey()
    const out3 = await hkdf(ikm2, null, info)
    expect(bytes(out3)).not.toEqual(bytes(out1))
  })

  it('10 — wrapMasterKeyWithPassphrase returns envelope, salt, params', async () => {
    const masterKey = generateMasterKey()
    const result = await wrapMasterKeyWithPassphrase(masterKey, 'test-passphrase', FAST_PARAMS)
    expect(result.envelope).toBeInstanceOf(Uint8Array)
    expect(result.envelope.length).toBeGreaterThan(30)
    expect(result.salt).toHaveLength(16)
    expect(result.params).toMatchObject(FAST_PARAMS)
  })

  it('11 — unwrapMasterKeyWithPassphrase round-trip', async () => {
    const masterKey = generateMasterKey()
    const { envelope, salt, params } = await wrapMasterKeyWithPassphrase(masterKey, 'correct-horse', FAST_PARAMS)
    const recovered = await unwrapMasterKeyWithPassphrase(envelope, 'correct-horse', salt, params)
    expect(bytes(recovered)).toEqual(bytes(masterKey))
  })

  it('12 — unwrapMasterKeyWithPassphrase with wrong passphrase throws', async () => {
    const masterKey = generateMasterKey()
    const { envelope, salt, params } = await wrapMasterKeyWithPassphrase(masterKey, 'right', FAST_PARAMS)
    await expect(unwrapMasterKeyWithPassphrase(envelope, 'wrong', salt, params)).rejects.toThrow()
  })

  it('13 — buildAsymmetricEnvelope / parseAsymmetricEnvelope round-trip preserves fields', () => {
    const alg = 'p256-ecdh-hkdf-aes256gcm-v1'
    const ephemeral = new Uint8Array(65).fill(0xAB)
    const nonce = generateNonce()
    const ct = new Uint8Array([5, 6, 7, 8])
    const env = buildAsymmetricEnvelope(alg, ephemeral, nonce, ct)
    const parsed = parseAsymmetricEnvelope(env)
    expect(parsed.algorithmId).toBe(alg)
    expect(bytes(parsed.ephemeralPubkeyBytes)).toEqual(bytes(ephemeral))
    expect(bytes(parsed.nonce)).toEqual(bytes(nonce))
    expect(bytes(parsed.ciphertextWithTag)).toEqual(bytes(ct))
  })

  it('15 — decryptStreamingContent round-trips a single-chunk streaming blob', async () => {
    const dek = generateDek()
    const plaintext = new TextEncoder().encode('streaming chunk test')
    // Nonce first byte is 0x41 ('A') — valid upload-id prefix byte, never 0x01.
    const nonce = new Uint8Array(12)
    nonce[0] = 0x41
    crypto.getRandomValues(nonce.subarray(1))
    const ct = await aesGcmEncryptWithAad(dek, nonce, nonce, plaintext)
    const streamingBlob = new Uint8Array(12 + ct.length)
    streamingBlob.set(nonce, 0)
    streamingBlob.set(ct, 12)
    const recovered = await decryptStreamingContent(streamingBlob, dek)
    expect(bytes(recovered)).toEqual(bytes(plaintext))
  })

  it('14 — wrapMasterKeyForDevice ECDH round-trip decrypts correctly', async () => {
    // Generate a device keypair (recipient)
    const deviceKeypair = await crypto.subtle.generateKey(
      { name: 'ECDH', namedCurve: 'P-256' }, true, ['deriveBits'],
    )
    const spki = new Uint8Array(await crypto.subtle.exportKey('spki', deviceKeypair.publicKey))

    const masterKey = generateMasterKey()
    const envelope = await wrapMasterKeyForDevice(masterKey, spki)

    // Decrypt: parse the asymmetric envelope, ECDH with device private key
    const { ephemeralPubkeyBytes, nonce, ciphertextWithTag } = parseAsymmetricEnvelope(envelope)

    const ephemeralPubKey = await crypto.subtle.importKey(
      'raw', ephemeralPubkeyBytes, { name: 'ECDH', namedCurve: 'P-256' }, false, [],
    )
    const sharedSecret = new Uint8Array(
      await crypto.subtle.deriveBits({ name: 'ECDH', public: ephemeralPubKey }, deviceKeypair.privateKey, 256),
    )
    const { hkdf: hkdfFn, aesGcmDecrypt: dec } = await import('../crypto/vaultCrypto')
    const kek = await hkdfFn(sharedSecret, null, new TextEncoder().encode('heirlooms-v1'))
    const recovered = await dec(kek, nonce, ciphertextWithTag)
    expect(bytes(recovered)).toEqual(bytes(masterKey))
  })
})
